### ISSUE TITLE
What are "Application Services" and how do we make them in Rails?

### ISSUE THREAD

**Issue by timfoley** [ga-wdi-exercises/project3#288]
I'm reading through [this article (Angular, What Goes Where](http://demisx.github.io/angularjs/2014/09/14/angular-what-goes-where.html) posted on the ngResources lesson plan, and I'm not sure what he means when he talks about Application Services in Rails (not Angular services).

![image](https://cloud.githubusercontent.com/assets/2591447/19860777/746b86c0-9f60-11e6-84ea-8ca893e28ee6.png)

> Keeping view logic in application controllers makes them eventually fat. Recommend placing it in the view service instead and have controllers simply call these view service methods and process their responses.

What is a view service, to pick one, and how do I work with it?

**Comment by jshawl** [comment]
Wait, what? What problem are you trying to solve?

**Comment by jshawl** [comment]
I'm not familiar with the term "view service"

**Comment by timfoley** [comment]
At a high level, we're having a heck of a time deciding where to put different parts of logic (does it go in Rails or Angular? If Rails, should it go in the controller, model methods, etc.).

This article seemed like a good explainer, but without knowing what he means by application services, I'm not sure where I should be putting certain logic if not in controllers.

**Comment by jshawl** [comment]
ah i see - can you give me a ELI5 definition? I could make a recommendation!

**Comment by timfoley** [comment]
### Our idea:

Recourse: a shared-bookmarks manager for web dev resources (find a great article on flexbox? Post it on Recourse!). 

### How We're Structuring Recourse:
- Use Rails to store our data and handle routes and views and then serve up data through our API (because it's got a lot in common with a classic CRUD app, so might as well take advantage of Rails routing) 
- Use Angular to display, sort and filter the data 

### Stuff We're Not Sure Where to Put:

We have a many-to-many relationship set up for both Favorites and Up/Downvotes. 

Neither really needs their own views (if someone wants to see their favorited posts, just select that filter option on the Post index).

Where should we put logic such as a method to toggle the "favorite" status of a particular post? Would that go in the PostsIndexControllerFunction in Angular? Would it go in the Posts controller in Rails? The models?

**Comment by jshawl** [comment]
Maybe a combination of all of the above?

angular: provides a physical button that lets the user click a favorite button. Angular controller sends post request to rails api indicating something is being favorited.

rails: has the endpoint for accepting the request?

not seeing an opprotunity for a model method here...

does that help?

**Comment by nolds9** [comment]
Ultimately its up to you, but as a general rule of thumb you will need some API endpoint on Rails tied to a controller action that will do the necessary create / update on the backend, then just some Angular view code for a button / controller code to make an Ajax call like Jesse suggested above...

**Comment by timfoley** [comment]
Hmmm interesting! 

So we might not even want to set up factories in angular that are doing CRUD because of the extra logic involved. Should angular just send 'post' requests back to our API signaling what needs to get done, and then we let our rails controllers handle validation, `@post.new(postParams)`, etc.?

**Comment by nolds9** [comment]
yes, depends on how your application is set up - I would still set up factories for each model that Im going to make traditional CRUD actions on, then just have separate API calls for the non-Restful routes like favorites...

**Comment by timfoley** [comment]
That makes sense, and brings me back to the question of where to put these things. 

I feel a bit shaky on my own understanding of what a non-crud API call would look like (where/how it gets fired from the front-end, where/how to set up our back-end to accept the call and act accordingly)

**Comment by nolds9** [comment]
Thats really because there is no one right answer, but in general you will need:
1. A route on the server side - some `url` that your front-end  will hit, you can define this yourself
2. That route needs to map to some controller action, again up to you which controller / action
3. Inside that controller action, it would take the request from the front end - perform the necessary application logic, then either send back an updated response in JSON, or redirect the user somewhere

On the front-end, really the only important things are which url you are hitting, what type of request it should be, and what data are you sending along with the request...the server should handle the heavy-lifting / logic.

Hope that helps

**Comment by timfoley** [comment]
Awesome. This is super helpful. That's the approach we'll take. 

Assuming we set up a route  for POST requests to the `/posts/:id/favorite` path connected to `Posts#favorite`, would it be a better or worse idea to fire those requests using `$.ajax ...` vs. something to do with factories? I've re-read the resources lesson a few times, but don't understand much about the role of factories beyond basic CRUD (i.e. I'm still confused about how we got "update" methods to work by adding that one line).

This is probably yet another question with a ton of possible answers, though. I appreciate all your responses so far!

**Comment by nolds9** [comment]
You have options, in short - totally fine use case for `$.ajax`.

When we set up a factory in Angular with `$resource`, that is a method that creates a wrapper around the 4 Restfull AJAX calls for the passed in endpoint.

To do one off requests that do not follow the rest convention, its totally fine to make a jQuery Ajax call like you suggested, or use angular's built in `$http` service to make requests https://docs.angularjs.org/api/ng/service/$http

**Comment by timfoley** [comment]
`$http` is exactly what I was looking for there. Perfect, thanks Nick!
